### PR TITLE
OCPBUGS-29126: deploymentcontroller: use builder pattern

### PR DESCRIPTION
**Pull request by bertinatto**
This is specially useful for controllers that don't want certain conditions.

CC @openshift/storage

### CODE REVIEW

**Comment by bertinatto** [comment]
/hold
for manual testing

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 1]
```diff
@@ -249,3 +309,12 @@ func isProgressing(deployment *appsv1.Deployment) (bool, string) {
 	}
 	return false, ""
 }
+
+func containsCondition(conditions []string, condition string) bool {
```
Comment (dusk125) [inline pkg/operator/deploymentcontroller/deployment_controller.go:313]: nit: could use slices.Contains from standard library

Comment (bertinatto) [reply pkg/operator/deploymentcontroller/deployment_controller.go:313]: Replaced

**Comment by bertinatto** [comment]
/hold cancel
tested at https://github.com/openshift/cluster-storage-operator/pull/456

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 2]
```diff
@@ -70,33 +74,84 @@ func NewDeploymentController(
...
+	c.optionalManifestHooks = hooks
+	return c
+}
+
+func (c *DeploymentController) WithDeploymentHooks(hooks ...DeploymentHookFunc) *DeploymentController {
+	c.optionalDeploymentHooks = hooks
+	return c
+}
 
+func (c *DeploymentController) WithConditions(conditions ...string) *DeploymentController {
```
Comment (deads2k) [inline pkg/operator/deploymentcontroller/deployment_controller.go:131]: missing doc.

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 3]
```diff
@@ -151,42 +206,48 @@ func (c *DeploymentController) syncManaged(ctx context.Context, opSpec *opv1.Ope
...
+			return nil
+		},
 	}
 
-	if ok, msg := isProgressing(deployment); ok {
-		progressingCondition.Status = opv1.ConditionTrue
-		progressingCondition.Message = msg
-		progressingCondition.Reason = "Deploying"
+	// Set Available Condition
+	if slices.Contains(c.conditions, opv1.OperatorStatusTypeAvailable) {
```
Comment (deads2k) [inline pkg/operator/deploymentcontroller/deployment_controller.go:218]: looks like `c.conditions` is a whitelist.  If it's a whitelist, error if anything unexpected is present.

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 4]
```diff
@@ -151,42 +206,48 @@ func (c *DeploymentController) syncManaged(ctx context.Context, opSpec *opv1.Ope
...
+		}
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(availableCondition))
 	}
 
-	updateStatusFn := func(newStatus *opv1.OperatorStatus) error {
-		// TODO: set ObservedGeneration (the last stable generation change we dealt with)
-		resourcemerge.SetDeploymentGeneration(&newStatus.Generations, deployment)
-		return nil
+	// Set Progressing Condition
+	if slices.Contains(c.conditions, opv1.OperatorStatusTypeProgressing) {
```
Comment (deads2k) [inline pkg/operator/deploymentcontroller/deployment_controller.go:234]: old code always  had these, why make them "sometimes"?

Comment (bertinatto) [reply pkg/operator/deploymentcontroller/deployment_controller.go:262]: The idea was to have the possibility to make every condition optional. `NewDeploymentController` still always has this by default.

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 5]
```diff
@@ -70,36 +78,111 @@ func NewDeploymentController(
...
+				c.conditions = append(c.conditions, condition)
+			}
+		}
+	}
+	return c
+}
 
-	return factory.New().WithInformers(
+// ToController converts the DeploymentController into a factory.Controller.
+func (c *DeploymentController) ToController() factory.Controller {
```
Comment (deads2k) [inline pkg/operator/deploymentcontroller/deployment_controller.go:163]: This could return an error so people would knwo they broke it.

Comment (bertinatto) [reply pkg/operator/deploymentcontroller/deployment_controller.go]: Nice! Done.

**Pull request merged**